remaining:
```
FAILED tests/test_general.py::test_runtime_handling[pickle] - assert 3 == -1
 +  where 3 = <function test_runtime_handling.<locals>.fn_minus at 0x106041510>(1, 2)
 +    where <function test_runtime_handling.<locals>.fn_minus at 0x106041510> = <function cachier.<locals>._cachier_decorator at 0x106041240>(<function test_runtime_handling.<locals>.fn_minus at 0x106040a60>)
FAILED tests/test_general.py::test_partial_handling - assert 3 == -1
 +  where 3 = <function cachier.<locals>._cachier_decorator.<locals>.func_wrapper at 0x106042440>()
 +    where <function cachier.<locals>._cachier_decorator.<locals>.func_wrapper at 0x106042440> = <function cachier.<locals>._cachier_decorator at 0x1060423b0>(functools.partial(<function test_partial_handling.<locals>.fn_minus at 0x106042290>, 1))
```

--- a/src/cachier/cores/pickle.py
+++ b/src/cachier/cores/pickle.py
@@ -83,6 +83,7 @@ class _PickleCore(_BaseCore):
         self.separate_files = _update_with_defaults(
             separate_files, "separate_files"
         )
+        self._cache_loaded_fpath = ""
 
     @property
     def cache_fname(self) -> str:
@@ -114,6 +115,7 @@ class _PickleCore(_BaseCore):
         try:
             with portalocker.Lock(self.cache_fpath, mode="rb") as cf:
                 cache = pickle.load(cf)  # noqa: S301
+            self._cache_loaded_fpath = str(self.cache_fpath)
         except (FileNotFoundError, EOFError):
             cache = {}
         return {
@@ -122,11 +124,14 @@ class _PickleCore(_BaseCore):
         }
 
     def get_cache_dict(self, reload: bool = False) -> Dict[str, CacheEntry]:
+        if self._cache_loaded_fpath != self.cache_fpath:
+            # force reload if the cache file has changed
+            # this change is dies to using different wrapped function
+            reload = True
         if self._cache_dict and not (self.reload or reload):
             return self._cache_dict
         with self.lock:
-            if not self._cache_dict:
-                self._cache_dict = self._load_cache_dict()
+            self._cache_dict = self._load_cache_dict()
         return self._cache_dict
 
     def _load_cache_by_key(

--- a/src/cachier/cores/pickle.py
+++ b/src/cachier/cores/pickle.py
@@ -9,7 +9,7 @@
 import os
 import pickle  # for local caching
 from datetime import datetime
-from typing import Any, Dict, Mapping, Optional, Tuple, Union
+from typing import Any, Dict, Optional, Tuple, Union
 
 import portalocker  # to lock on pickle cache IO
 from watchdog.events import PatternMatchingEventHandler

--- a/src/cachier/cores/pickle.py
+++ b/src/cachier/cores/pickle.py
@@ -83,7 +83,7 @@ class _PickleCore(_BaseCore):
         self.separate_files = _update_with_defaults(
             separate_files, "separate_files"
         )
-        self._cache_loaded_fpath = ""
+        self._cache_used_fpath = ""
 
     @property
     def cache_fname(self) -> str:
@@ -115,7 +115,7 @@ class _PickleCore(_BaseCore):
         try:
             with portalocker.Lock(self.cache_fpath, mode="rb") as cf:
                 cache = pickle.load(cf)  # noqa: S301
-            self._cache_loaded_fpath = str(self.cache_fpath)
+            self._cache_used_fpath = str(self.cache_fpath)
         except (FileNotFoundError, EOFError):
             cache = {}
         return {
@@ -124,7 +124,7 @@ class _PickleCore(_BaseCore):
         }
 
     def get_cache_dict(self, reload: bool = False) -> Dict[str, CacheEntry]:
-        if self._cache_loaded_fpath != self.cache_fpath:
+        if self._cache_used_fpath != self.cache_fpath:
             # force reload if the cache file has changed
             # this change is dies to using different wrapped function
             reload = True
@@ -181,6 +181,8 @@ class _PickleCore(_BaseCore):
         with self.lock:
             with portalocker.Lock(fpath, mode="wb") as cf:
                 pickle.dump(cache, cf, protocol=4)
+            if not separate_file_key:
+                self._cache_used_fpath = str(self.cache_fpath)
             # the same as check for separate_file, but changed for typing
             if isinstance(cache, dict):
                 self._cache_dict = cache

--- a/tests/test_general.py
+++ b/tests/test_general.py
@@ -443,14 +443,14 @@ def test_runtime_handling(tmpdir, backend):
     assert count_p == count_m == 0
 
     for fn, expected in [(fn_plus, 3), (fn_minus, -1)]:
-        assert cachier_(fn)(1, 2) == expected
-        assert cachier_(fn)(a=1, b=2) == expected
+        assert cachier_(fn)(1, 2) == expected, f"for {fn.__name__} inline"
+        assert cachier_(fn)(a=1, b=2) == expected, f"for {fn.__name__} inline"
     assert count_p == 1
     assert count_m == 1
 
     for fn, expected in [(fn_plus, 5), (fn_minus, 1)]:
-        assert cachier_(fn)(3, 2) == expected
-        assert cachier_(fn)(a=3, b=2) == expected
+        assert cachier_(fn)(3, 2) == expected, f"for {fn.__name__} inline"
+        assert cachier_(fn)(a=3, b=2) == expected, f"for {fn.__name__} inline"
     assert count_p == 2
     assert count_m == 2
 
@@ -473,22 +473,24 @@ def test_partial_handling(tmpdir):
 
     for fn, expected in [(fn_plus, 3), (fn_minus, -1)]:
         dummy_ = functools.partial(fn, 1)
-        assert cachier_(dummy_)() == expected
+        assert cachier_(dummy_)() == expected, f"for {fn.__name__} wrapped"
 
         dummy_ = functools.partial(fn, 1)
-        assert cachier_(dummy_)(2) == expected
+        assert cachier_(dummy_)(2) == expected, f"for {fn.__name__} wrapped"
 
         dummy_ = functools.partial(fn, a=1)
-        assert cachier_(dummy_)() == expected
+        assert cachier_(dummy_)() == expected, f"for {fn.__name__} wrapped"
 
         dummy_ = functools.partial(fn, b=2)
-        assert cachier_(dummy_)(1) == expected
+        assert cachier_(dummy_)(1) == expected, f"for {fn.__name__} wrapped"
 
         dummy_ = functools.partial(fn, b=2)
-        assert cachier_(dummy_)(1, b=2) == expected
+        assert (
+            cachier_(dummy_)(1, b=2) == expected
+        ), f"for {fn.__name__} wrapped"
 
-        assert cachier_(fn)(1, 2) == expected
-        assert cachier_(fn)(a=1, b=2) == expected
+        assert cachier_(fn)(1, 2) == expected, f"for {fn.__name__} inline"
+        assert cachier_(fn)(a=1, b=2) == expected, f"for {fn.__name__} inline"
 
     assert count_p == 1
     assert count_m == 1

